File: MoreHighlight/src/main/java/xyz/wingio/plugins/morehighlight/HeaderNode.java
package xyz.wingio.plugins.morehighlight;

import androidx.core.graphics.ColorUtils;

import com.aliucord.PluginManager;
import com.discord.simpleast.core.node.Node;
import com.discord.utilities.textprocessing.*;

import android.text.SpannableStringBuilder;
import android.text.style.*;

public class HeaderNode<MessageRenderContext> extends Node<MessageRenderContext> {
  int headerSize;

  public HeaderNode(int headerSize){
    super();
    this.headerSize = headerSize;
  }

  @Override
  public void render(SpannableStringBuilder builder, MessageRenderContext renderContext) {
    int length = builder.length();
    for (Node n: getChildren()) {
      n.render(builder, renderContext);
    }
    
    float scaleFactor = PluginManager.plugins.get("MoreHighlight").settings.getFloat("header_size_scale", 1.0f);
  
    switch (headerSize) {
      case 1:
        builder.setSpan(new RelativeSizeSpan(2.0f * scaleFactor), length, builder.length(), SpannableStringBuilder.SPAN_EXCLUSIVE_EXCLUSIVE);
        break;
      case 2:
        builder.setSpan(new RelativeSizeSpan(1.5f * scaleFactor), length, builder.length(), SpannableStringBuilder.SPAN_EXCLUSIVE_EXCLUSIVE);
        break;
      case 3:
        builder.setSpan(new RelativeSizeSpan(1.25f * scaleFactor), length, builder.length(), SpannableStringBuilder.SPAN_EXCLUSIVE_EXCLUSIVE);
        break;
    }
    builder.setSpan(new StyleSpan(1), length, builder.length(), SpannableStringBuilder.SPAN_EXCLUSIVE_EXCLUSIVE);
  }
}

File: MoreHighlight/src/main/java/xyz/wingio/plugins/morehighlight/node/HeaderNode.kt
package xyz.wingio.plugins.morehighlight.node

import android.graphics.Typeface
import android.text.SpannableStringBuilder
import android.text.style.RelativeSizeSpan
import android.text.style.StyleSpan
import com.aliucord.PluginManager
import com.discord.simpleast.core.node.Node
import xyz.wingio.plugins.morehighlight.MoreHighlight

class HeaderNode<MessageRenderContext>(
    private val headerSize: Int
): Node.a<MessageRenderContext>() {

    override fun render(builder: SpannableStringBuilder, renderContext: MessageRenderContext) {
        val length = builder.length
        super.render(builder, renderContext)

        val scaleFactor = PluginManager.plugins["MoreHighlight"]!!.settings.getFloat(MoreHighlight.PREF_HEADER_SIZE, 1.0f)
        val proportion = when (headerSize) {
            1 -> 2.0f * scaleFactor
            2 -> 1.5f * scaleFactor
            3 -> 1.25f * scaleFactor
            else -> 1f
        }

        builder.setSpan(RelativeSizeSpan(proportion), length, builder.length, SpannableStringBuilder.SPAN_EXCLUSIVE_EXCLUSIVE)
        builder.setSpan(StyleSpan(Typeface.BOLD), length, builder.length, SpannableStringBuilder.SPAN_EXCLUSIVE_EXCLUSIVE)
    }

}
